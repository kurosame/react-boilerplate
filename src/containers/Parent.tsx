import Child from '@/components/Child'
import React from 'react'
import styled from 'styled-components'

const Parent = () => (
  <Div>
    Parent
    <Child />
  </Div>
)

const Div = styled.div`
  color: white;
  background-color: green;
`

export default Parent
