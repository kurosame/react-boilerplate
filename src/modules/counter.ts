import { Reducer } from 'redux'
import {
  Action,
  ActionFunctionAny,
  createActions,
  handleActions
} from 'redux-actions'

// Action types
export const ADD_COUNT = 'ADD_COUNT'
export const ADD_SAGA_COUNT = 'ADD_SAGA_COUNT'
export const GET_SAGA_COUNT = 'GET_SAGA_COUNT'

export interface CounterActions {
  addCount: ActionFunctionAny<Action<CounterState>>
  addSagaCount: ActionFunctionAny<Action<CounterState>>
  getSagaCount: ActionFunctionAny<Action<CounterState>>
}

export interface CounterState {
  count: number
  sagaCount: number
}

const initialState: CounterState = {
  count: 0,
  sagaCount: 0
}

export const { addCount, addSagaCount, getSagaCount } = createActions<
  CounterState
>(ADD_COUNT, ADD_SAGA_COUNT, GET_SAGA_COUNT)

export const counter: Reducer<
  CounterState,
  Action<CounterState>
> = handleActions(
  {
    [ADD_COUNT]: (state: CounterState): CounterState => {
      return { ...state, count: state.count + 1 }
    },
    [ADD_SAGA_COUNT]: (
      state: CounterState,
      action: Action<CounterState>
    ): CounterState => {
      return {
        ...state,
        sagaCount: state.sagaCount + action.payload.sagaCount
      }
    }
  },
  initialState
)
