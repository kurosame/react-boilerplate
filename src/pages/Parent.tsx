import React from 'react'
import styled from 'styled-components'
import Child from '@/components/Child'

const Parent = (): JSX.Element => (
  <Div>
    Parent
    <Child />
  </Div>
)

const Div = styled.div`
  color: white;
  background-color: green;
`

export default Parent
