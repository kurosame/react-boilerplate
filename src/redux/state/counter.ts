export type State = {
  count: number
  axiosCount: number
  asyncAwaitCount: number
}
