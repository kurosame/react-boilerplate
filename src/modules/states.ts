import { CounterState } from '@/modules/counter.ts'

export interface States {
  counter: CounterState
}
