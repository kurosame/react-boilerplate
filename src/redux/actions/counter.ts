import { createActions, ActionFunctionAny, Action } from 'redux-actions'
import { ADD_VALUE } from '@/redux/types'
import { State } from '@/redux/state/counter'

export type Actions = {
  addValue: ActionFunctionAny<Action<{ count: number }>>
}

export const { addValue } = createActions({
  [ADD_VALUE]: (): { count: number } => {
    return { count: 1 }
  }
})
