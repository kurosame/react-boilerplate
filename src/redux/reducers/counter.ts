import { IState } from '@/redux/state/counter'
import { ADD_VALUE, SAGA_SAMPLE } from '@/redux/types'
import { Action, handleActions, Reducer } from 'redux-actions'

const initialState: IState = {
  count: 0,
  sagaCount: 0
}

export const counter: Reducer<IState, IState> = handleActions(
  {
    [ADD_VALUE]: (state: IState): IState => {
      return { ...state, count: state.count + 1 }
    },
    [SAGA_SAMPLE]: (state: IState, action: Action<IState>): IState => {
      return {
        ...state,
        sagaCount:
          state.sagaCount + (action.payload ? action.payload.sagaCount : 0)
      }
    }
  },
  initialState
)
