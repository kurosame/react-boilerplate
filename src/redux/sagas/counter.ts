import { put, take, call, fork } from 'redux-saga/effects'
import axios from 'axios'
import { SAGA_SAMPLE } from '@/redux/types'
import { sagaSample } from '@/redux/actions/counter'

function* getSagaSample() {
  while (true) {
    yield take(SAGA_SAMPLE)
    const sagaCount: number = yield call(() =>
      axios
        .get('/api')
        .then(res => res.data.sagaCount)
        .catch(() => console.error('SAGA_SAMPLE API response error'))
    )
    yield put(sagaSample({ sagaCount }))
  }
}

export default function* rootSaga() {
  yield fork(getSagaSample)
}
