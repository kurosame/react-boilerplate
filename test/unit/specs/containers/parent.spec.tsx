import { mount } from 'enzyme'
import React from 'react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import Parent from '@/containers/Parent.tsx'
import { CounterState } from '@/modules/counter.ts'
import { States } from '@/modules/states.ts'

const state: { counter: CounterState } = {
  counter: { count: 12, sagaCount: 34 }
}
const states: States = { counter: state.counter }

const wrapper = mount(
  <Provider store={configureStore<States>()(states)}>
    <Parent />
  </Provider>
)

test('Match the snapshot', () => {
  wrapper
    .find('[data-test="add-count"]')
    .first()
    .simulate('click')
  wrapper
    .find('[data-test="add-count"]')
    .last()
    .simulate('click')

  expect(wrapper.html()).toMatchSnapshot()
})
