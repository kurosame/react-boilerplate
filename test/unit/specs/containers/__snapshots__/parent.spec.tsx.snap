// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match the snapshot 1`] = `"<div class=\\"sc-AykKC exfSga\\"><span data-test=\\"count\\">12</span><button data-test=\\"add-count\\">ADD</button></div><div class=\\"sc-AykKC exfSga\\"><span data-test=\\"count\\">34</span><button data-test=\\"add-count\\">ADD</button></div>"`;
